### PR TITLE
Stricter default arguments

### LINKED ISSUE #1453: Optional arguments spec
Hi,

Optional arguments are a good way to provide some kind of static overloading in a clean way.

However, one thing that bothers me is the order of the parameters.

Say we have this example

~~~nit
class A
    fun foo(x: String, y: Int, z: nullable Int) is abstract
end

var a = new A
a.foo("Hi", 10)
a.foo("Hia", 1, 10)
~~~

All good here, we see what is going on, the first call is with the null z, the other with z = 10

But, what if we choose to define a service like this:

~~~nit
redef class A
    fun bar(x: nullable String, y: String, z: nullable String, a: nullable Int) is abstract
end

var a = new A
a.bar("",a=0)
a.bar("","",a=0)
~~~

The second call is quite strange imho, the spec does say that the x="" and y="" variant is called, however, is this really what we meant ?

This gets harder with constructors:

~~~nit
class A
    var x: nullable String
    var y: String
end

class B super A
    var z: nullable String
end

var a = new B("", "")
~~~

What is it that we're instanciating ? Is it x and y ? y and z ? It seems unclear to me, especially if the specialization is done in a submodule, one can get surprised easily

I think that, to KISSize the spec, we should change the way optional arguments are parsed, namely:

* Mandatory arguments first, in linearization order
* Facultative arguments last, in linearization order

This spec might limit the surprises with the spec by placing mandatory arguments before facultative ones, which, in constructors specifically, might be surprising.

I know that the actual spec is pretty clear on what I'm supposed to get with what I wrote above, but I feel like I'd be less surprised if mandatory arguments were always first and optional arguments after that.
I believe this would lead to a more uniform way of calling methods and constructors.

### PR DESCRIPTION
**Pull request by privat**
This change the way that default arguments are handled.

Now, defaults arguments can only follow the used one.
This improve the readability of calls and the association between calls and declarations thus solve a lot of issues of users.

~~~
fun foo(a: nullable Int, b: Int, c: nullable Int) do ...
foo(null,2,null) # OK
foo(null,2) # equivalent
foo(2) # now refused! but previously accepted as equivalent.
~~~

Only exception: the last parameter of an assignment method is always the last argument

~~~
fun foo=(a: nullable Int, b: Int, c: nullable Int, d: nullable Int) do ...
foo(null,2,null) = 0 # OK
foo(null,2) = 0 # equivalent
foo(2) = 0 # now refused! but previously accepted as equivalent
~~~

No specific black magic is added to automatic constructor.
Therefore an optional constructor parameter in a class can becore mandatory in a subclass if a new mandatory attribute is introduced.

~~~
class A
   var a: nullable Int
end
class B
  super A
  var b: Int
end
var a1 = new A(null) # OK
var a2 = new A # equivalent
var b1 = new B(null, 2) # OK
var b2 = new B(2) # now refused! but previously accepted as equivalent
~~~

This issue only required some small adaptation in existing piece of code: nitdoc, sexp, serialization and dom. The latter is PRized independently in #1616 but included here for jenkins.
Most of these changes (first commits) are in fact bugfixes or make the code cleaner so this is an argument in favor of this new stricter specification.

This PR has also the advantage of simplifying the model as the whole policy of the default argument is moved to the typing analysis: it is now just a pure calling convention.

Close #1453

### CODE REVIEW

**Comment by R4PaSs** [comment]
:shipit:

**Comment by Morriar** [comment]
If I'm correct, varargs also have to be in last position, so you canot use varargs and default arguments? What about default varargs?

File: lib/dom/parser.nit [hunk 1]
```diff
@@ -170,70 +170,67 @@ class XMLProcessor
...
-		var idst = pos
+		if pos >= srclen then return new XMLError(st_loc, "Invalid start of prolog")
 		var tag_name = parse_tag_name(['<', '>'])
 		var c = src[pos]
-		if c == '<' or c == '>' then return new XMLError(location=st_loc ,"Unexpected character `{c}` in prolog declaration")
+		if c == '<' or c == '>' then return new XMLError(st_loc ,"Unexpected character `{c}` in prolog declaration")
 		if tag_name == "xml" then
 			var args = parse_args(['?'])
 			for i in args do
-				if i isa BadXMLAttribute then return new XMLError(location = i.location, i.name)
```
Comment (Morriar) [inline lib/dom/parser.nit]: I like how named arguments where supposed to lighten the code :)

File: src/semantize/typing.nit [hunk 2]
```diff
@@ -414,26 +414,30 @@ private class TypeVisitor
...
 
+
 		#debug("CALL {unsafe_type}.{msignature}")
 
 		# Associate each parameter to a position in the arguments
 		var map = new SignatureMap
 
-		var setted = args.length - msignature.min_arity
+		# Special case for the isolated last argument
+		# TODO: reify this method characteristic (where? the param, the signature, the method?)
```
Comment (Morriar) [inline src/semantize/typing.nit]: characteristics*

File: src/semantize/typing.nit [hunk 3]
```diff
@@ -456,10 +456,12 @@ private class TypeVisitor
 				return null
 			end
 			map.map[idx] = i
-			setted -= 1
 			e.mtype = self.visit_expr_subtype(e.n_expr, param.mtype)
 		end
 
+		# Number of minimum mandatory remaining parameters
```
Comment (Morriar) [inline src/semantize/typing.nit]: If these have to do with default arguments, you should say it.

**Comment by Morriar** [comment]
+1

I like to see nit in the right way again.

**Comment by privat** [comment]
@Morriar varargs did not change: it is either vararg or default arguments.

File: lib/dom/parser.nit (hunk 1 above)
Comment (privat) [inline lib/dom/parser.nit]: I assumed a minimum amount of sanity from library developers; I was wrong.

**Pull request merged**